src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -111,34 +106,70 @@ public static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
             Contracts.CheckValue(env, nameof(env));
             env.CheckValue(ctx, nameof(ctx));
             ctx.CheckAtModel(GetVersionInfo());
+
             // *** Binary format ***
+            // int: indicator for frozen models
             // stream: tensorFlow model.
             // int: number of input columns
             // for each input column
             //   int: id of int column name
             // int: number of output columns
             // for each output column
             //   int: id of output column name
-            byte[] modelBytes = null;
-            if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
-                throw env.ExceptDecode();
-            var numInputs = ctx.Reader.ReadInt32();
-            env.CheckDecode(numInputs > 0);
-            string[] inputs = new string[numInputs];
-            for (int j = 0; j < inputs.Length; j++)
-                inputs[j] = ctx.LoadNonEmptyString();
+            bool isFrozen = true;
+            bool isNonFrozenModelSupported = ctx.Header.ModelVerReadable >= 0x00010003;
+            if (isNonFrozenModelSupported)
+                isFrozen = ctx.Reader.ReadBoolByte();
 
-            bool isMultiOutput = ctx.Header.ModelVerReadable >= 0x00010002;
-            var numOutputs = 1;
-            if (isMultiOutput)
-                numOutputs = ctx.Reader.ReadInt32();
+            if (isFrozen)
+            {
+                byte[] modelBytes = null;
+                if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
+                    throw env.ExceptDecode();
 
-            env.CheckDecode(numOutputs > 0);
-            var outputs = new string[numOutputs];
-            for (int j = 0; j < outputs.Length; j++)
-                outputs[j] = ctx.LoadNonEmptyString();
+                var io = ModelInputsOutputs(env, ctx);
```

Comment:
>var io = ModelInputsOutputs(env, ctx); [](start = 16, length = 38)

This line can be moved before the if.

Reply:
If we move it before the if (line127)  wont it break the  order in which it was saved ?  

(The model inputs/outputs come only after  we are done  reading the stream contents)

---
In reply to: [217559827](https://github.com/dotnet/machinelearning/pull/853#discussion_r217559827) [](ancestors = 217559827)

Reply:
Since the model is saved in a separate stream, I think it doesn't matter. The order is important only for what ends up in the Model.key file.

---
In reply to: [217571319](https://github.com/dotnet/machinelearning/pull/853#discussion_r217571319) [](ancestors = 217571319,217559827)